python/bin/p1_print.py

```diff
@@ -145,179 +150,104 @@ def print_message(header, contents, one_line=False):
-            if index_file is not None:
-                if next_message_count == len(index_file):
-                    break
-
-                # If we're reading all messages, read the next index entry. Otherwise, read the entry for the next
-                # message we're interested in.
-                if message_indices is None:
-                    message_index = next_message_count
+    for header, message, data, offset_bytes in reader:
+        bytes_decoded += len(data)
+
+        # Update the data summary in summary mode, or print the message contents otherwise.
+        if options.summary:
+            p1_time = message.get_p1_time()
+            if p1_time is not None:
+                if first_p1_time_sec is None:
+                    first_p1_time_sec = float(p1_time)
+                    last_p1_time_sec = float(p1_time)
+                    newest_p1_time = p1_time
+                else:
+                    if p1_time < newest_p1_time:
+                        _logger.warning('P1 time restart detected after %s.' % str(newest_p1_time))
+                    last_p1_time_sec = max(last_p1_time_sec, float(p1_time))
+                    newest_p1_time = p1_time
+
+            system_time_sec = message.get_system_time_sec()
+            if system_time_sec is not None:
+                if first_system_time_sec is None:
+                    first_system_time_sec = system_time_sec
+                    last_system_time_sec = system_time_sec
+                    newest_system_time_sec = system_time_sec
                 else:
-                    message_index = message_indices[next_message_count]
-                next_message_count += 1
-
-                # Determine the offset to the message of interest. The index file doesn't store message sizes, so we
-                # just deserialize the message header to figure out how much to read.
-                offset_bytes = index_file.offset[message_index]
-
-                f.seek(offset_bytes, io.SEEK_SET)
-                header_data = f.read(MessageHeader.calcsize())
-                header = MessageHeader()
-                header.unpack(buffer=header_data, warn_on_unrecognized=False)
-
-                payload_data = f.read(header.payload_size_bytes)
-                data = header_data + payload_data
-                bytes_read = f.tell()
-            # Otherwise, if we do not have an index file, read the next chunk of data and process all of it.
+                    if system_time_sec < newest_system_time_sec:
+                        _logger.warning('System time restart detected after %s.' %
+                                        system_time_to_str(newest_system_time_sec, is_seconds=True))
+                    last_system_time_sec = max(last_system_time_sec, system_time_sec)
+                    newest_system_time_sec = system_time_sec
+
+            total_messages += 1
+            if header.message_type not in message_stats:
+                message_stats[header.message_type] = {
```

Comment:
Its a bit nicer to use a  defaultdict for this sort of pattern.

Reply:
Sure, good idea.